addons/sale_project/tests/test_sale_project.py

```diff
@@ -652,3 +652,82 @@ def test_quick_create_sol(self):
+                "standard_price": 30,
+                "list_price": 90,
+                "type": "service",
+                "service_tracking": "task_global_project",
+                "invoice_policy": "order",
+                "uom_id": uom_hour.id,
+                "uom_po_id": uom_hour.id,
+                "project_id": multi_company_project.id,
+            })
+
+        def create_and_confirm_sale_order(company, product):
+            sale_order = SaleOrder.with_company(company).create({
+                "partner_id": will_smith.id,
+                "order_line": [
+                    Command.create({
+                        "product_id": product.id,
+                        "product_uom_qty": 10,
+                    }),
+                    Command.create({
+                        "product_id": product.id,
+                        "product_uom_qty": 10,
+                    }),
+                ],
+            })
+            sale_order.with_company(company).action_confirm()
+            return sale_order
+
+        company_a, company_b = self.env['res.company'].create([
+            {"name": "Company A"},
+            {"name": "Company B"},
+        ])
+
+        product_a = create_product(company_a)
+        sale_order_a = create_and_confirm_sale_order(company_a, product_a)
+
+        product_b = create_product(company_b)
+        sale_order_b = create_and_confirm_sale_order(company_b, product_b)
+
+        sale_order_per_company = {
+            company_a.id: sale_order_a,
+            company_b.id: sale_order_b,
+        }
+        for company in [company_a, company_b]:
+            self.assert_multi_company_project_sols_are_accessible(multi_company_project, company,
+                                                                  sale_order_per_company)
+
+    def assert_multi_company_project_sols_are_accessible(self, multi_company_project, company, sale_order_per_company):
+        # Counts reflect all sale orders / sale order lines.
+        self.assertEqual(
+            multi_company_project.with_company(company).sale_order_count, len(sale_order_per_company),
+            "Expected all sale orders to be counted by project")
+        self.assertEqual(
+            multi_company_project.with_company(company).sale_order_line_count,
+            len([sol.id for _, sale_order in sale_order_per_company.items() for sol in sale_order.order_line]),
+            "Expected all sale order lines lines to be counted by project")
+
+        # Viewing sale order lines only returns those that belong to the current company.
+        id_domain = next(filter(lambda triple: triple[0] == 'id',
+                                multi_company_project.with_company(company).action_view_sols()['domain']))
+        self.assertEqual(
+            id_domain[2],
+            sale_order_per_company[company.id].order_line.ids,
+            "Expected sale order lines that belong to the current company to be returned")
```

Comment:
No need to check the domain, actually, we can give many ids as we want, the ir rules will do the job to filter the records that satisfy that domain to only get the ones the user can read.

The issue in the screenshot you mentioned in one of your comments was in fact a traceback inside the execution of `action_view_sos` method and not a traceback because we load too much records in the view. 🙂

Reply:
Hmm, haven't noticed this, thanks for pointing that out. But I think that both actions need to be filtered as no SOLs that belong to other companies need to be returned for a certain company, don't you think so? Is there a better way to do this than to filter after `_fetch_sale_order_items`?

Reply:
No, because the user could select more than one company and so you will filter too much. Moreover, there is no issue if we don't filter so I don't know why you really want to filter that since we don't actually filter in the count.

And so since we don't have any issue during the process of `action_view_sols` method, I would prefer let as it is and if the user would like to select all the companies and then click on the stat button then he will really see all the SOLs linked to that project. 🙂 

(same opinion for `action_view_sos` method) 🙂

Reply:
Oh, I didn't actually know you can select multiple companies and the access would be their union. Thanks for pointing that out!